feat(pagination): Implement pagination for tag pages

shelton-xiaoteng-ma opened the pull request:
Implement pagination for tag pages using searchParams instead of dynamic router to minimize changes

- Added pagination support to tag pages.
- Replaced dynamic router with searchParams for pagination, reducing the number of changes to existing routing logic.
- Updated navigation links to include pagination parameters via searchParams.
- Ensured backwards compatibility with previous URL structures while introducing pagination.

timlrx commented:
Thanks, there is a small bug introduced with the change. The main `tags` page still shows the "Next" button as active despite it being the last page: https://tailwind-nextjs-starter-blog-git-fork-s-a5ce5a-timlrxs-projects.vercel.app/blog?page=2

shelton-xiaoteng-ma commented:
Ok, I have solved it. 
This bug is caused by using the same `ListLayout` from `@/layouts/ListLayoutWithTags`, now I have refactored `/blog`, deleted `/blog/[page]/page.tsx`, and made sure `/blog` and `/tags/next-js` uses the same pagination logic

timlrx commented:
Thanks, the `not found` logic seems to be different. For e.g. https://tailwind-nextjs-starter-blog-cxns6sbn4-timlrxs-projects.vercel.app/blog?page=3 directs to page 1 despite there not being a valid page 3. It displays a 404 error for the tag routes.

shelton-xiaoteng-ma commented:
The URL you are using is outdated, I saw the new deployment URL: `https://tailwind-nextjs-starter-blog-nvzeomgx1-timlrxs-projects.vercel.app/`, so `https://tailwind-nextjs-starter-blog-nvzeomgx1-timlrxs-projects.vercel.app/blog?page=3` shows `Not Found` as expected
<img width="1209" alt="Screenshot 2025-01-07 at 15 07 27" src="https://github.com/user-attachments/assets/d643afae-0e00-46fe-9b4a-dc03526269cb" />

timlrx commented:
I get a bunch of build errors for the static version of the site:
```
Error occurred prerendering page "/tags/markdown". Read more: https://nextjs.org/docs/messages/prerender-error
Error: Route /tags/[tag] with `dynamic = "error"` couldn't be rendered statically because it used `await searchParams`, `searchParams.then`, or similar. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering
Export encountered an error on /tags/[tag]/page: /tags/markdown, exiting the build.
Error occurred prerendering page "/tags/images". Read more: https://nextjs.org/docs/messages/prerender-error
Error: Route /tags/[tag] with `dynamic = "error"` couldn't be rendered statically because it used `await searchParams`, `searchParams.then`, or similar. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering
Export encountered an error on /tags/[tag]/page: /tags/images, exiting the build.
```
I think the inclusion of `searchParams` changes it from statically rendered to dynamic rendering. Do you have any work arounds that would still allow it to be statically rendered?

shelton-xiaoteng-ma commented:
Sorry for the new error. I am not familiar with this.
ChatGPT says that you can use "precompute all possible values", I think it works for tags, but for pages, Maybe `Use useSearchParams and handle logic on the client`. 
I will try other options later.
